core/src/impl/KokkosExp_IterateTileGPU.hpp

```diff
@@ -319,168 +246,73 @@ struct DeviceIterateTile<4, PolicyType, Functor, MaxGridSize, Tag> {
+      const index_type thread_id_1 = threadIdx.x / m_policy.m_tile[0];
+      const index_type thread_id_0 = threadIdx.x % m_policy.m_tile[0];
+
+      for (index_type offset_3 = start_3;
+           offset_3 < static_cast<index_type>(m_policy.m_upper[3]);
+           offset_3 += stride_3) {
+        for (index_type offset_2 = start_2;
+             offset_2 < static_cast<index_type>(m_policy.m_upper[2]);
+             offset_2 += stride_2) {
+          // Reconstruct tile 0 and tile 1 from blockIdx.x and threadIdx.x
+          for (index_type tile_x = blockIdx.x; tile_x < max_tiles_01;
+               tile_x += gridDim.x) {
+            const index_type tile_1 = tile_x / m_policy.m_tile_end[0];
+            const index_type tile_0 = tile_x % m_policy.m_tile_end[0];
+
+            const index_type offset_1 =
+                tile_1 * m_policy.m_tile[1] + thread_id_1 + m_policy.m_lower[1];
+            const index_type offset_0 =
+                tile_0 * m_policy.m_tile[0] + thread_id_0 + m_policy.m_lower[0];
+
+            if (offset_1 < static_cast<index_type>(m_policy.m_upper[1]) &&
+                offset_0 < static_cast<index_type>(m_policy.m_upper[0])) {
+              Impl::_tag_invoke<Tag>(m_func, offset_0, offset_1, offset_2,
+                                     offset_3);
             }
           }
         }
       }
-    }
-    // LR
-    else {
-      // number of tiles for dimension 0
-      const index_type temp0 = m_policy.m_tile_end[0];
-      // number of tiles for dimension 1
-      const index_type temp1 = m_policy.m_tile_end[1];
-
-      // number of virtual blocks for dimension 1
-      const index_type numbl1 =
-          Kokkos::min(temp1, static_cast<index_type>(m_max_grid_size[0]));
-      // number of virtual blocks for dimension 0
-      const index_type numbl0 =
-          (temp0 * temp1 > static_cast<index_type>(m_max_grid_size[0])
-               ? static_cast<index_type>(m_max_grid_size[0]) / numbl1
-               : Kokkos::min(temp0,
-                             static_cast<index_type>(m_max_grid_size[0])));
-
-      // first virtual block index for dimension 0
-      const index_type tile_id0 = static_cast<index_type>(blockIdx.x) / numbl1;
-      // first virtual block index for dimension 1
-      const index_type tile_id1 = static_cast<index_type>(blockIdx.x) % numbl1;
-
-      // virtual thread index for dimension 0
-      const index_type thr_id0 =
-          static_cast<index_type>(threadIdx.x) / m_policy.m_tile[1];
-      // virtual thread index for dimension 1
-      const index_type thr_id1 =
-          static_cast<index_type>(threadIdx.x) % m_policy.m_tile[1];
-
-      // iterate over virtual blocks for dimension 0
-      for (index_type i = tile_id0; i < m_policy.m_tile_end[0]; i += numbl0) {
-        // compute index for dimension 0
+
+    } else {  // Iterate::Right
+
+      const index_type thread_id_0 = threadIdx.x / m_policy.m_tile[1];
```

Comment:
If we have the default tile size `(16, 2, 2, 2)` for `Iterate::Left` and `(2, 2, 2, 16)` for `Iterate::Right`, this gives the asymmetrical behavior.
We lunch the kernel with `(32, 2, 2)` for `Iterate::Left` and `(4, 2, 16)` for `Iterate::Right`.